contents/Documentation/FP concepts.playground/Pages/Glossary.xcplaygroundpage/Contents.swift

```diff
@@ -10,5 +10,85 @@
+ # A
+
+ ### Ad-hoc polymorphism
+
+ **Ad-hoc polymorphism** is a type of polymorphism achieved by having functions with the same name but different signatures and/or different number of arguments. In Functional Programming, ad-hoc polymorphism is achieved with type classes that act as constraints added to type parameters in parametrically polymorphic types or functions. Ad-hoc polymorphism is also known as function overloading.
+
+ # F
+
+ ### Function
+
+ A function is a piece of code that has the following properties:
+
+ - **Total**: it must provide an output for every input.
+ - **Deterministic**: for a given input, the function always returns the same output.
+ - **Pure**: the evaluation of the function does not cause any other effects besides computing the output.
+
+ Notice that Swift does not enforce any of these properties in their definition. Swift functions that have these properties are generally referred as **pure functions**, whereas the ones that do not have some of them are called **impure functions**.
+
+ ### Functional Programming
+
+ **Functional Programming** is a programming paradigm that treats computation as the evaluation of mathematical functions and avoids changing state and mutable data.
+
+ # I
+
+ ### Instance (of a type class)
+
+ An **instance** of a type class is the implementation of such type class for a given type. They are usually done in Swift using the `extension` functionality in the language.
+
+ # K
+
+ ### Kind
+
+ A **kind** is a group of types. As values are grouped into types, types are grouped into kinds. The notion of kind lets us conceptualize type constructors that receive a number of type arguments and abstract over them. This generalization is known as **higher kinded types** and enable generic polymorphic programs. Swift does not support this feature natively, but it is simulated in Bow.
+
+ # M
+
+ ### Memoization
+
+ **Memoization** is an optimization technique that transforms expensive computational function calls into searches in a lookup table by caching previous calls.
+
+ # P
+
+ ### Parametric polymorphism
+
+ **Parametric polymorphism** is a type of polymorphism that allows to specify generic types to create functions and/or other types. These generic types are called *type parameters*.
+
+ ### Procedure
+
+ A **procedure** is a Swift function that is not total, deterministic and/or pure, as required in the mathematical definition of function. Generally, it is referred to as an **impure function**.
+
+ # R
+
+ ### Referential Transparency
+
+ **Referential transparency** is a property of a function that allows it to be replaced by the result of its execution without altering the overall behavior of the program.
+
+ # T
+
+ ### Type
+
+ A **type** is a set of values that helps us conceptualize data and restrict the possible values that functions can accept as input or provide as output. Types can be finite (like `Bool`, with two values: `true` and `false`) or infinite (like `String`).
```

Comment:
```suggestion
 A **type** is a set of values that help us conceptualize data and restrict the possible values that functions can accept as input or provide as output. Types can be finite (like `Bool`, with two values: `true` and `false`) or infinite (like `String`).
```

Reply:
set -> helps

Reply:
Ops, you cheating! haha, yep, you are right